File: workloads/parboil/lbm/hip/lbm.cpp
// lbm.cpp -- HIP benchmark for Parboil Lattice-Boltzmann Method (D3Q19)
// Kernel: Each thread updates one lattice cell by streaming + collision
// using D3Q19 model with two arrays (ping-pong) for src/dst.

#include "bench_common_hip.h"

// D3Q19 velocity directions
#define Q 19

// D3Q19 weights
__constant__ float d_w[Q] = {
    1.0f/3.0f,                                          // 0: rest
    1.0f/18.0f, 1.0f/18.0f, 1.0f/18.0f,                // 1-3: +x,+y,+z
    1.0f/18.0f, 1.0f/18.0f, 1.0f/18.0f,                // 4-6: -x,-y,-z
    1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f,    // 7-10: xy diags
    1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f,    // 11-14: xz diags
    1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f     // 15-18: yz diags
};

// D3Q19 velocity vectors (ex, ey, ez)
__constant__ int d_ex[Q] = { 0,  1, 0, 0, -1,  0,  0,  1,-1, 1,-1,  1,-1, 1,-1,  0, 0, 0, 0};
__constant__ int d_ey[Q] = { 0,  0, 1, 0,  0, -1,  0,  1, 1,-1,-1,  0, 0, 0, 0,  1,-1, 1,-1};
__constant__ int d_ez[Q] = { 0,  0, 0, 1,  0,  0, -1,  0, 0, 0, 0,  1, 1,-1,-1,  1, 1,-1,-1};

// Opposite direction indices for bounce-back
__constant__ int d_opp[Q] = { 0,  4, 5, 6,  1,  2,  3, 10, 9, 8, 7, 14,13,12,11, 18,17,16,15};

// Stream + Collide kernel
// src: source distribution, dst: destination distribution
// N: grid dimension (NxNxN), omega: relaxation parameter
__global__ void lbm_stream_collide(float *src, float *dst, int N, float omega) {
    int idx = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
    int total = N * N * N;
    if (idx >= total) return;

    // Convert linear index to 3D
    int z = idx / (N * N);
    int y = (idx / N) % N;
    int x = idx % N;

    // Streaming: gather from neighbors
    float f[Q];
    for (int q = 0; q < Q; q++) {
        // Source cell for direction q (streaming step: pull from neighbor)
        int sx = (x - d_ex[q] + N) % N;
        int sy = (y - d_ey[q] + N) % N;
        int sz = (z - d_ez[q] + N) % N;
        int src_idx = (sz * N * N + sy * N + sx) * Q + q;
        f[q] = src[src_idx];
    }

    // Compute macroscopic quantities: density and velocity
    float rho = 0.0f;
    float ux = 0.0f, uy = 0.0f, uz = 0.0f;
    for (int q = 0; q < Q; q++) {
        rho += f[q];
        ux += f[q] * d_ex[q];
        uy += f[q] * d_ey[q];
        uz += f[q] * d_ez[q];
    }

    // Avoid division by zero
    float inv_rho = (rho > 1e-10f) ? (1.0f / rho) : 0.0f;
    ux *= inv_rho;
    uy *= inv_rho;
    uz *= inv_rho;

    // Collision step: BGK relaxation toward equilibrium
    float u_sq = ux * ux + uy * uy + uz * uz;
    for (int q = 0; q < Q; q++) {
        float eu = (float)d_ex[q] * ux + (float)d_ey[q] * uy + (float)d_ez[q] * uz;
        float feq = d_w[q] * rho * (1.0f + 3.0f * eu + 4.5f * eu * eu - 1.5f * u_sq);
        f[q] = f[q] * (1.0f - omega) + feq * omega;
    }

    // Write results to destination
    for (int q = 0; q < Q; q++) {
        dst[idx * Q + q] = f[q];
    }
}

int main(int argc, char **argv) {
    int iterations = parseIterations(argc, argv);
    int grid_dim = parseIntParam(argc, argv, "--grid_dim", 32);
    int num_timesteps = parseIntParam(argc, argv, "--num_timesteps", 10);
    int omega_x100 = parseIntParam(argc, argv, "--omega_x100", 160);
    int block_size = parseIntParam(argc, argv, "--block_size", 256);

    float omega = omega_x100 / 100.0f;
    int total_cells = grid_dim * grid_dim * grid_dim;
    size_t lattice_size = (size_t)total_cells * Q * sizeof(float);

    // Allocate host memory
    float *h_lattice = (float *)malloc(lattice_size);

    // Initialize with equilibrium distribution (rho=1, u=0)
    srand(42);
    for (int i = 0; i < total_cells; i++) {
        // Small random perturbation around equilibrium
        float rho = 1.0f + 0.01f * ((rand() % 200) - 100) / 100.0f;
        // D3Q19 weights for equilibrium at rest
        float w[Q] = {
            1.0f/3.0f,
            1.0f/18.0f, 1.0f/18.0f, 1.0f/18.0f,
            1.0f/18.0f, 1.0f/18.0f, 1.0f/18.0f,
            1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f,
            1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f,
            1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f, 1.0f/36.0f
        };
        for (int q = 0; q < Q; q++) {
            h_lattice[i * Q + q] = rho * w[q];
        }
    }

    // Device allocations (two arrays for ping-pong)
    float *d_src, *d_dst;
    HIP_CHECK(hipMalloc(&d_src, lattice_size));
    HIP_CHECK(hipMalloc(&d_dst, lattice_size));

    HIP_CHECK(hipMemcpy(d_src, h_lattice, lattice_size,
                        hipMemcpyHostToDevice));

    dim3 block(block_size);
    dim3 grid((total_cells + block_size - 1) / block_size);

    char problemSize[64];
    snprintf(problemSize, sizeof(problemSize), "%d", grid_dim);

    printCSVHeader();

    // Benchmark: run num_timesteps of stream+collide per iteration
    BenchResult r = runBenchmark(
        "lbm_stream_collide", problemSize, iterations, [&]() {
            for (int t = 0; t < num_timesteps; t++) {
                lbm_stream_collide<<<grid, block>>>(d_src, d_dst, grid_dim, omega);
                // Swap src and dst (ping-pong)
                float *tmp = d_src;
                d_src = d_dst;
                d_dst = tmp;
            }
        });
    printCSVRow(r);

    // Cleanup
    HIP_CHECK(hipFree(d_src));
    HIP_CHECK(hipFree(d_dst));
    free(h_lattice);

    return 0;
}

File: amd/benchmarks/parboil/lbm/lbm.go
// Package lbm implements the Parboil LBM (Lattice Boltzmann Method, D3Q19)
// benchmark, ported from sarchlab/gpu_benchmarks (tier2/parboil_lbm) for the
// MGPUSim MI300A (CDNA3 / gfx942) model.
//
// It simulates fluid flow on a regular NxNxN 3D grid using the D3Q19 lattice.
// A fused collide-stream kernel applies the BGK collision operator and streams
// the post-collision distributions to neighbor cells, with bounce-back
// boundary conditions at the grid boundaries. The host iterates the kernel for
// a fixed number of timesteps, swapping the source/destination distribution
// buffers each step.
//
// The kernel binary is compiled for gfx942 only (see native/), so the
// benchmark must be run with `-arch cdna3` (the MI300A configuration). The
// D3Q19 velocity set, weights and opposite tables live as __constant__ arrays
// baked into the code object, and the kernel uses a constant block size, so no
// hidden ABI arguments are emitted (kernarg_segment_size = 32).
package lbm

import (
	"log"
	"math"

	// embed hsaco files
	_ "embed"

	"github.com/sarchlab/mgpusim/v5/amd/arch"
	"github.com/sarchlab/mgpusim/v5/amd/driver"
	"github.com/sarchlab/mgpusim/v5/amd/insts"
)

// Q is the number of discrete velocities in the D3Q19 lattice.
const Q = 19

// blockSize is the compile-time block dimension baked into the kernel.
const blockSize = 128

// D3Q19 velocity set, weights and opposite-direction table. These mirror the
// __constant__ tables baked into the kernel and are used by the CPU reference.
var (
	hEx = [Q]int{0, 1, -1, 0, 0, 0, 0, 1, -1, 1, -1, 1, -1, 1, -1, 0, 0, 0, 0}
	hEy = [Q]int{0, 0, 0, 1, -1, 0, 0, 1, 1, -1, -1, 0, 0, 0, 0, 1, -1, 1, -1}
	hEz = [Q]int{0, 0, 0, 0, 0, 1, -1, 0, 0, 0, 0, 1, 1, -1, -1, 1, 1, -1, -1}

	hW = [Q]float32{
		1.0 / 3.0,
		1.0 / 18.0, 1.0 / 18.0, 1.0 / 18.0,
		1.0 / 18.0, 1.0 / 18.0, 1.0 / 18.0,
		1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0,
		1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0,
		1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0, 1.0 / 36.0,
	}

	hOpp = [Q]int{0, 2, 1, 4, 3, 6, 5, 10, 9, 8, 7, 14, 13, 12, 11, 18, 17, 16, 15}
)

// KernelArgs defines the kernel arguments for the gfx942 (CDNA3) kernel.
//
// The layout is verified against the compiled kernel's AMDGPU metadata
// (kernarg_segment_size = 32): two 8-byte global_buffer pointers followed by
// four 4-byte by_value scalars, packed with no padding (mgpusim serializes
// args with binary.Write, which does not insert alignment padding). The kernel
// uses constant block dimensions and __constant__ tables only, so no hidden
// ABI arguments are emitted.
type KernelArgs struct {
	FSrc  driver.Ptr // offset 0
	FDst  driver.Ptr // offset 8
	Nx    int32      // offset 16
	Ny    int32      // offset 20
	Nz    int32      // offset 24
	Omega float32    // offset 28
}

// Benchmark defines the LBM benchmark.
type Benchmark struct {
	driver  *driver.Driver
	context *driver.Context
	queue   *driver.CommandQueue
	hsaco   *insts.KernelCodeObject
	gpus    []int

	Arch         arch.Type
	GridDim      int     // N for the NxNxN grid
	NumTimesteps int     // number of collide-stream iterations
	Tau          float32 // relaxation time; omega = 1/tau

	hF    []float32 // host initial distribution (Q*N)
	gFSrc driver.Ptr
	gFDst driver.Ptr

	useUnifiedMemory bool
}

//go:embed kernels_gfx942.hsaco
var cdna3HSACOBytes []byte

// NewBenchmark returns a new LBM benchmark.
func NewBenchmark(driver *driver.Driver) *Benchmark {
	b := new(Benchmark)

	b.driver = driver
	b.context = b.driver.Init()
	b.queue = driver.CreateCommandQueue(b.context)

	return b
}

func (b *Benchmark) loadProgram() {
	b.hsaco = insts.LoadKernelCodeObjectFromBytes(
		cdna3HSACOBytes, "lbm_collide_stream_kernel")
	if b.hsaco == nil {
		log.Panic("Failed to load kernel binary")
	}
}

// SelectGPU selects the GPUs to run on. LBM uses a single GPU.
func (b *Benchmark) SelectGPU(gpus []int) {
	b.gpus = gpus
}

// SetUnifiedMemory requests the use of unified memory.
func (b *Benchmark) SetUnifiedMemory() {
	b.useUnifiedMemory = true
}

// Run runs the benchmark.
func (b *Benchmark) Run() {
	if b.Arch != arch.CDNA3 {
		log.Panic("the parboil lbm benchmark ships only a gfx942 " +
			"kernel; run with -arch cdna3 -gpu mi300a")
	}

	b.loadProgram()

	b.driver.SelectGPU(b.context, b.gpus[0])
	b.initMem()
	b.exec()
}

// lcgRand reproduces the host LCG PRNG from the original benchmark.
func lcgRand(state *uint32) uint32 {
	*state = *state*1664525 + 1013904223
	return *state
}

// randFloat reproduces the host rand_float helper.
func randFloat(state *uint32, lo, hi float32) float32 {
	return lo + float32(lcgRand(state)&0xFFFF)/65535.0*(hi-lo)
}

func (b *Benchmark) initMem() {
	if b.GridDim <= 0 {
		b.GridDim = 16
	}
	if b.NumTimesteps <= 0 {
		b.NumTimesteps = 4
	}
	if b.Tau <= 0 {
		b.Tau = 0.7
	}

	n := b.GridDim
	numCells := n * n * n

	// Deterministic equilibrium initialization, identical to the HIP host.
	b.hF = make([]float32, Q*numCells)
	seed := uint32(42)
	for idx := 0; idx < numCells; idx++ {
		rho := 1.0 + 0.001*randFloat(&seed, -1.0, 1.0)
		for q := 0; q < Q; q++ {
			b.hF[q*numCells+idx] = hW[q] * rho
		}
	}

	zeros := make([]float32, Q*numCells)

	if b.useUnifiedMemory {
		b.gFSrc = b.driver.AllocateUnifiedMemory(
			b.context, uint64(Q*numCells*4))
		b.gFDst = b.driver.AllocateUnifiedMemory(
			b.context, uint64(Q*numCells*4))
	} else {
		b.gFSrc = b.driver.AllocateMemory(
			b.context, uint64(Q*numCells*4))
		b.gFDst = b.driver.AllocateMemory(
			b.context, uint64(Q*numCells*4))
	}

	b.driver.MemCopyH2D(b.context, b.gFSrc, b.hF)
	b.driver.MemCopyH2D(b.context, b.gFDst, zeros)
}

func (b *Benchmark) exec() {
	n := b.GridDim
	numCells := n * n * n
	omega := float32(1.0) / b.Tau

	gridSize := uint32((numCells + blockSize - 1) / blockSize)
	globalX := gridSize * blockSize

	src := b.gFSrc
	dst := b.gFDst

	for t := 0; t < b.NumTimesteps; t++ {
		args := KernelArgs{
			FSrc:  src,
			FDst:  dst,
			Nx:    int32(n),
			Ny:    int32(n),
			Nz:    int32(n),
			Omega: omega,
		}

		b.driver.EnqueueLaunchKernel(
			b.queue,
			b.hsaco,
			[3]uint32{globalX, 1, 1},
			[3]uint16{blockSize, 1, 1},
			&args,
		)
		b.driver.DrainCommandQueue(b.queue)

		src, dst = dst, src
	}

	// After the loop, the final result lives in src (it was last written-to
	// destination before the final swap).
	b.gFSrc = src
	b.gFDst = dst
}

// cpuCollideStream runs one collide-stream timestep on the CPU, identical to
// the device kernel.
func cpuCollideStream(fSrc, fDst []float32, n int, omega float32) {
	numCells := n * n * n
	nx, ny, nz := n, n, n

	for idx := 0; idx < numCells; idx++ {
		iz := idx / (nx * ny)
		iy := (idx / nx) % ny
		ix := idx % nx

		isBoundary := ix == 0 || ix == nx-1 ||
			iy == 0 || iy == ny-1 ||
			iz == 0 || iz == nz-1

		var f [Q]float32
		for q := 0; q < Q; q++ {
			f[q] = fSrc[q*numCells+idx]
		}

		var rho, ux, uy, uz float32
		for q := 0; q < Q; q++ {
			rho += f[q]
			ux += f[q] * float32(hEx[q])
			uy += f[q] * float32(hEy[q])
			uz += f[q] * float32(hEz[q])
		}
		invRho := 1.0 / float32(math.Max(float64(rho), 1e-10))
		ux *= invRho
		uy *= invRho
		uz *= invRho

		u2 := ux*ux + uy*uy + uz*uz
		var fPost [Q]float32
		for q := 0; q < Q; q++ {
			eu := float32(hEx[q])*ux + float32(hEy[q])*uy + float32(hEz[q])*uz
			fEq := hW[q] * rho * (1.0 + 3.0*eu + 4.5*eu*eu - 1.5*u2)
			fPost[q] = f[q] + omega*(fEq-f[q])
		}

		for q := 0; q < Q; q++ {
			nxi := ix + hEx[q]
			nyi := iy + hEy[q]
			nzi := iz + hEz[q]

			if isBoundary {
				fDst[hOpp[q]*numCells+idx] = fPost[q]
			} else if nxi >= 0 && nxi < nx && nyi >= 0 && nyi < ny &&
				nzi >= 0 && nzi < nz {
				nidx := nzi*nx*ny + nyi*nx + nxi
				fDst[q*numCells+nidx] = fPost[q]
			} else {
				fDst[hOpp[q]*numCells+idx] = fPost[q]
			}
		}
	}
}

// Verify checks the GPU result against a CPU reference computation.
func (b *Benchmark) Verify() {
	n := b.GridDim
	numCells := n * n * n
	omega := float32(1.0) / b.Tau

	gpuF := make([]float32, Q*numCells)
	b.driver.MemCopyD2H(b.context, gpuF, b.gFSrc)

	// CPU reference: replay the exact same timesteps with buffer swapping.
	src := make([]float32, Q*numCells)
	dst := make([]float32, Q*numCells)
	copy(src, b.hF)
	for i := range dst {
		dst[i] = 0
	}

	for t := 0; t < b.NumTimesteps; t++ {
		cpuCollideStream(src, dst, n, omega)
		src, dst = dst, src
	}
	// src now holds the final result.

	for i := 0; i < Q*numCells; i++ {
		ref := float64(src[i])
		got := float64(gpuF[i])

		denom := math.Abs(ref)
		if denom < 1.0 {
			denom = 1.0
		}
		if math.Abs(ref-got)/denom > 1e-3 {
			q := i / numCells
			cell := i % numCells
			log.Fatalf("At q=%d cell=%d, expected %f, but got %f.\n",
				q, cell, ref, got)
		}
	}

	log.Printf("Passed!\n")
}
